pyomo/contrib/solver/base.py

```diff
@@ -0,0 +1,557 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2024
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import abc
+import enum
+from typing import Sequence, Dict, Optional, Mapping, NoReturn, List, Tuple
+import os
+
+from pyomo.core.base.constraint import _GeneralConstraintData
+from pyomo.core.base.var import _GeneralVarData
+from pyomo.core.base.param import _ParamData
+from pyomo.core.base.block import _BlockData
+from pyomo.core.base.objective import _GeneralObjectiveData
+from pyomo.common.config import document_kwargs_from_configdict
+from pyomo.common.errors import ApplicationError
+from pyomo.common.deprecation import deprecation_warning
+from pyomo.opt.results.results_ import SolverResults as LegacySolverResults
+from pyomo.opt.results.solution import Solution as LegacySolution
+from pyomo.core.kernel.objective import minimize
+from pyomo.core.base import SymbolMap
+from pyomo.core.base.label import NumericLabeler
+from pyomo.core.staleflag import StaleFlagManager
+from pyomo.contrib.solver.config import SolverConfig, PersistentSolverConfig
+from pyomo.contrib.solver.util import get_objective
+from pyomo.contrib.solver.results import (
+    Results,
+    legacy_solver_status_map,
+    legacy_termination_condition_map,
+    legacy_solution_status_map,
+)
+
+
+class SolverBase(abc.ABC):
+    """
+    This base class defines the methods required for all solvers:
+        - available: Determines whether the solver is able to be run, combining both whether it can be found on the system and if the license is valid.
+        - solve: The main method of every solver
+        - version: The version of the solver
+        - is_persistent: Set to false for all non-persistent solvers.
+
+    Additionally, solvers should have a :attr:`config<SolverBase.config>` attribute that
+    inherits from one of :class:`SolverConfig<pyomo.contrib.solver.config.SolverConfig>`,
+    :class:`BranchAndBoundConfig<pyomo.contrib.solver.config.BranchAndBoundConfig>`,
+    :class:`PersistentSolverConfig<pyomo.contrib.solver.config.PersistentSolverConfig>`, or
+    :class:`PersistentBranchAndBoundConfig<pyomo.contrib.solver.config.PersistentBranchAndBoundConfig>`.
+    """
+
+    CONFIG = SolverConfig()
```

Comment:
I don't disagree; however, the benefit of having this declared is in autodoc generation. Because this is declared, I can use it to generate docs for `SolverBase` that show the default config items that should be on _all_ solvers.

Reply:
OK - but then we need to both document that that *must* be overwritten by ALL solver implementations, AND we need to set up a test that iterates over the solver classes in the solver factory and checks that they were all overwritten

Reply:
That is a good point @jsiirola. I'll work on a test now.

Reply:
Test added.